File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> count;

        for (int num : nums) {
            count[num] = count[num] + 1;
        }

        // initialize array size n to iterate frequency k times
        // this makes O(n) complexity possible
        vector<vector<int>> freq(nums.size() + 1);

        // set count as key and num as value
        for (const auto &c : count) {
            freq[c.second].push_back(c.first);
        }

        vector<int> ans;
        
        // iterate from back of freq array
        for (int i = freq.size() - 1; i > 0; --i) {
            for (int f : freq[i]) {
                ans.push_back(f);
                if (ans.size() == k) {
                    return ans;
                }
            }
        }

        return ans;
    }
};

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        if len(s) == 1 or len(s) == k:
            return len(s)
        ans = 0

        l, r = 0, 0
        maxf = 0
        count = {}

        while r < len(s):
            count[s[r]] = count.get(s[r], 0) + 1
            maxf = max(maxf, count[s[r]]) 

            while (r - l + 1) - maxf > k:
                count[s[l]] -= 1
                l += 1
            
            ans = max(ans, r - l + 1)
            r += 1

        return ans
